name: Update Code

on:
  issue_comment:
    types: [created, edited]


permissions:
  contents: write
  pull-requests: write
  issues: write

jobs:
  testing:
    name: Deploy Beta
    runs-on: ubuntu-latest
    if: ${{ contains(github.event.comment.body, 'test_branch') }}
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          token: ${{ secrets.API_TOKEN_GITHUB }}

      - name: Checkout Pull Request
        run: hub pr checkout ${{ github.event.issue.number }}
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

      - uses: ./.github/workflows/release_deploy.yml
        with:
          semver: buildNumber
          environment: dev
          beta: true
        secrets: inherit

      - uses: ./.github/workflows/integration_test.yml
        with:
          environment: dev
        secrets: inherit



  notify:
    needs: [ testing ]
    runs-on: ubuntu-latest
    name: Notify
    if: ${{ always() && contains(needs.*.result, 'failure') }}
    steps:
      - name: Notify if Failure
        uses: actions/github-script@v6.3.3
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          script: |
            var comments = await github.rest.issues.listComments({
                issue_number: context.issue.number,
                owner: context.repo.owner,
                repo: context.repo.repo
            });
            for (const comment of comments.data) {
              if (comment.body.includes('Test is failed. Please retry.')){
                github.rest.issues.deleteComment({
                  issue_number: context.issue.number,
                  owner: context.repo.owner,
                  repo: context.repo.repo,
                  comment_id: comment.id
                })
              }
            }
            github.rest.issues.createComment({
              issue_number: context.issue.number,
              owner: context.repo.owner,
              repo: context.repo.repo,
              body: 'Test is failed. Please retry.'
            })
            core.setFailed('Test is failed. Please retry.')

